core/src/main/java/org/apache/calcite/rel/rules/AggregateExpandDistinctAggregatesRule.java

```diff
@@ -151,6 +167,19 @@ public void onMatch(RelOptRuleCall call) {
       return;
     }
 
+    // If only one distinct aggregate and one or more non-distinct aggregates,
+    // we can generate multi-phase aggregates
+    if (distinctCount == 1                // one distinct aggregate
+            && !hasGroupingSets           // no grouping sets
+            && !hasUnsupportedAggFunc     // sum/min/max/count in non-distinct aggregate
+            && nonDistinctCount > 0       // one or more non-distinct aggregates
+            && aggregate.getGroupSet().size() > 0) { // at least one group-by key present
```

Comment:
Oh, I don't know exactly. Maybe you need to cast a nullable int to a
not-null int. Maybe you need a "case when x is null then 0 else x end".
Maybe you need to use sum0 rather than sum. But query returns the right
answer, doesn't it? So the information is there.

On Wed, Jun 15, 2016 at 6:15 PM, Gautam Parai <notifications@github.com>
wrote:

> In
> core/src/main/java/org/apache/calcite/rel/rules/AggregateExpandDistinctAggregatesRule.java
> <https://github.com/apache/calcite/pull/247#discussion_r67273712>:
>
> > @@ -151,6 +167,19 @@ public void onMatch(RelOptRuleCall call) {
> >        return;
> >      }
> >
> > +    // If only one distinct aggregate and one or more non-distinct aggregates,
> > +    // we can generate multi-phase aggregates
> > +    if (distinctCount == 1                // one distinct aggregate
> > +            && !hasGroupingSets           // no grouping sets
> > +            && !hasUnsupportedAggFunc     // sum/min/max/count in non-distinct aggregate
> > +            && nonDistinctCount > 0       // one or more non-distinct aggregates
> > +            && aggregate.getGroupSet().size() > 0) { // at least one group-by key present
>
> Sorry, I did not understand that. Could you please explain with the
> example above assuming you mean a SQL CAST?
> {source}
> select count(
>
>
>
>
>
> *), count(distinct deptno) from emp {source} would transform to {source}
> select sum(cnt), count(deptno) from (select count(*) as cnt, deptno
> from emp
> group by deptno)
> {source}
> but will fail nullability checks.
> What should the CAST form look like?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/calcite/pull/247/files/7cb85dbc5261ee65c66a16c94c85f464be37da62#r67273712>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/ABFAOXX7aRWcIzU6ioY5UoC7QVaMNBXqks5qMKPIgaJpZM4IzbCE>
> .
>